flips/20220725-cadence-borrowContract.md

```diff
@@ -0,0 +1,75 @@
+# borrowContract function for PublicAccount
+
+| Status        | Proposed    												  |
+:-------------- |:------------------------------------------------------------|
+| **FLIP #**    | [NNN](https://github.com/onflow/flow/pull/NNN)  	 		  |
+| **Author(s)** | Deniz Mert Edincik (deniz@edincik.com)  					  |
+|				| Austin Kline (austin@flowty.io)  							  | 
+| **Sponsor**   | 														      |
+| **Updated**   | 2022-07-25                                    			  |
+
+## Objective
+
+This proposed change will allow dynamic import of contracts from an address, by
+using a well known `borrow` dynamic.
+
+## Motivation
+
+There is a popular request, especially from people coming from EVM space, to
+have a way to import contracts from an address by name ( dynamic import
+functionality )
+
+Cadence lacks with dynamic imports, which leads to mass imports in some 
+contracts ( e.g. alchemy, find ) which also causes a lot of contracts to fail, 
+if one of the important contracts cease to exist. 
+
+## User Benefit
+
+- It makes it possible to use patterns that are common to Solidity developers 
+in Flow. 
+- It allows importing and using Contracts with name collision https://github.com
+/onflow/cadence/issues/1171
+- It will prevent single contract breaking the ecosystem
+
+## Design Proposal
+
+We suggest to add an new method to `PublicAccount` called `borrowContract`.
+
+```cadence 
+fun borrowContract<T: &Any>(name: String): T?
+```
+
+Exmaple usage will be:
+
+```cadence 
+if let contract = getAccount(0x1).borrowContract<&AnyResource{NonFungibleToken}>("ExampleNFT") {
```

Comment:
One of my concerns with this approach is that, it allows having multiple variables pointing to the same contract. And also makes contracts behave like other cadence 'values'.

I think this is something that has been avoided by the design of cadence: https://docs.onflow.org/cadence/language/contracts/
> Contracts are types. They are similar to composite types, but are stored differently than structs or resources and cannot be used as values, copied, or moved like resources or structs

Reply:
Yeah actually, I choose borrow exactly for this reason. ( multiple variables pointing to the same contract ) 

By using reference to the contract, we avoid copy and move technically.

Reply:
That's a good idea. Could you please add this reasoning to the proposal itself?